src/converse-mam-views.js

```diff
@@ -100,26 +113,34 @@ converse.plugins.add('converse-mam-views', {
                     return;
                 }
                 this.addSpinner();
-                _converse.api.archive.query(
-                    Object.assign({
-                        'groupchat': is_groupchat,
-                        'before': '', // Page backwards from the most recent message
-                        'max': _converse.archived_messages_page_size,
-                        'with': this.model.get('jid'),
-                    }, options),
-
-                    messages => { // Success
-                        this.clearSpinner();
-                        _.each(messages, message_handler);
-                    },
-                    e => { // Error
-                        this.clearSpinner();
-                        _converse.log(
-                            "Error or timeout while trying to fetch "+
-                            "archived messages", Strophe.LogLevel.ERROR);
-                        _converse.log(e, Strophe.LogLevel.ERROR);
-                    }
-                );
+
+                return new Promise((resolve, reject) => {
+                    _converse.api.archive.query(
+                        Object.assign({
+                            'groupchat': is_groupchat,
+                            'before': '', // Page backwards from the most recent message
+                            'max': _converse.archived_messages_page_size,
+                            'with': this.model.get('jid'),
+                        }, options),
```

Comment:
The next release of Converse is 5.0.0, so we can break the API.

I therefore think we should let `_converse.api.archive.query` return a Promise instead of accepting callback functions. That would make the code here much cleaner and nicer.

You won't need to do `return new Promise` and can instead use `await` statements.